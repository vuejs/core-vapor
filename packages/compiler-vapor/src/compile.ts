import {
  type CompilerOptions as BaseCompilerOptions,
  ErrorCodes,
  type RootNode,
  createCompilerError,
  defaultOnError,
  parse,
} from '@vue/compiler-dom'
import { extend, isString } from '@vue/shared'
import {
  type DirectiveTransform,
  type NodeTransform,
  transform,
} from './transform'
import { type VaporCodegenResult, generate } from './generate'
import { transformOnce } from './transforms/vOnce'
import { transformElement } from './transforms/transformElement'
import { transformVHtml } from './transforms/vHtml'
import { transformVText } from './transforms/vText'
import { transformVBind } from './transforms/vBind'
import { transformVOn } from './transforms/vOn'
import { transformVShow } from './transforms/vShow'
import { transformRef } from './transforms/transformRef'
import { transformInterpolation } from './transforms/transformInterpolation'
import type { HackOptions } from './ir'
import { transformVModel } from './transforms/vModel'
import { transformVIf } from './transforms/vIf'
import { transformVFor } from './transforms/vFor'

export type CompilerOptions = HackOptions<BaseCompilerOptions>

// TODO: copied from @vue/compiler-core
// code/AST -> IR -> JS codegen
export function compile(
  source: string | RootNode,
  options: CompilerOptions = {},
): VaporCodegenResult {
  const onError = options.onError || defaultOnError
  const isModuleMode = options.mode === 'module'
  /* istanbul ignore if */
  if (__BROWSER__) {
    if (options.prefixIdentifiers === true) {
      onError(createCompilerError(ErrorCodes.X_PREFIX_ID_NOT_SUPPORTED))
    } else if (isModuleMode) {
      onError(createCompilerError(ErrorCodes.X_MODULE_MODE_NOT_SUPPORTED))
    }
  }

  const prefixIdentifiers =
    !__BROWSER__ && (options.prefixIdentifiers === true || isModuleMode)

  // TODO scope id
  // if (options.scopeId && !isModuleMode) {
  //   onError(createCompilerError(ErrorCodes.X_SCOPE_ID_NOT_SUPPORTED))
  // }

  const resolvedOptions = extend({}, options, {
    prefixIdentifiers,
  })
  const ast = isString(source) ? parse(source, resolvedOptions) : source
  const [nodeTransforms, directiveTransforms] =
    getBaseTransformPreset(prefixIdentifiers)

  if (!__BROWSER__ && options.isTS) {
    const { expressionPlugins } = options
    if (!expressionPlugins || !expressionPlugins.includes('typescript')) {
      resolvedOptions.expressionPlugins = [
        ...(expressionPlugins || []),
        'typescript',
      ]
    }
  }

  const ir = transform(
    ast,
    extend({}, resolvedOptions, {
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {}, // user transforms
      ),
    }),
  )

  return generate(ir, resolvedOptions)
}

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>,
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean,
): TransformPreset {
  return [
    [
      transformOnce,
      transformRef,
      transformInterpolation,
      transformVIf,
      transformVFor,
      transformElement,
    ],
    {
      bind: transformVBind,
      on: transformVOn,
      html: transformVHtml,
      text: transformVText,
      show: transformVShow,
      model: transformVModel,
    },
  ]
}
