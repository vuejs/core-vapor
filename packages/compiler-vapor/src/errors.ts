import {
  CompilerError,
  SourceLocation,
  createCompilerError,
} from '@vue/compiler-dom'

export interface VaporCompilerError extends CompilerError {
  code: VaporErrorCodes
}

export function createVaporCompilerError(
  code: VaporErrorCodes,
  loc?: SourceLocation,
) {
  return createCompilerError(
    code,
    loc,
    __DEV__ || !__BROWSER__ ? VaporErrorMessages : undefined,
  ) as VaporCompilerError
}

export enum VaporErrorCodes {
  X_V_PLACEHOLDER = 100,
  __EXTEND_POINT__,
}

export const VaporErrorMessages: Record<VaporErrorCodes, string> = {
  [VaporErrorCodes.X_V_PLACEHOLDER]: `[placeholder]`,

  // just to fulfill types
  [VaporErrorCodes.__EXTEND_POINT__]: ``,
}
