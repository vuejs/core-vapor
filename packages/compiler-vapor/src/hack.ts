import type { Prettify } from '@vue/shared'
import type { DirectiveTransform, NodeTransform } from './transform'

type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> &
  Pick<U, Extract<keyof U, keyof T>>

export type HackOptions<T> = Prettify<
  Overwrite<
    T,
    {
      nodeTransforms?: NodeTransform[]
      directiveTransforms?: Record<string, DirectiveTransform | undefined>
    }
  >
>
