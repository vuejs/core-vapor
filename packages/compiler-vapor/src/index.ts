export { parse } from '@vue/compiler-dom'
export * from './transform'
export * from './generate'
export {
  wrapTemplate,
  compile,
  type CompilerOptions,
  type TransformPreset,
} from './compile'
export * from './ir'
export * from './errors'
export { transformElement } from './transforms/transformElement'
export { transformText } from './transforms/transformText'
export { transformVBind } from './transforms/vBind'
export { transformVHtml } from './transforms/vHtml'
export { transformVOn } from './transforms/vOn'
export { transformOnce } from './transforms/vOnce'
export { transformVShow } from './transforms/vShow'
export { transformVText } from './transforms/vText'
export { transformVIf } from './transforms/vIf'
export { transformVFor } from './transforms/vFor'
export { transformVModel } from './transforms/vModel'
