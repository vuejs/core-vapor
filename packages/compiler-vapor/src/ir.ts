import type {
  CompoundExpressionNode,
  DirectiveNode,
  RootNode,
  SimpleExpressionNode,
  SourceLocation,
} from '@vue/compiler-dom'
import type { Prettify } from '@vue/shared'
import type { DirectiveTransform, NodeTransform } from './transform'

export enum IRNodeTypes {
  ROOT,
  TEMPLATE_FACTORY,
  FRAGMENT_FACTORY,

  SET_PROP,
  SET_TEXT,
  SET_EVENT,
  SET_HTML,

  INSERT_NODE,
  PREPEND_NODE,
  APPEND_NODE,
  CREATE_TEXT_NODE,

  WITH_DIRECTIVE,
}

export interface BaseIRNode {
  type: IRNodeTypes
  loc: SourceLocation
}

// TODO refactor
export type VaporHelper = keyof typeof import('../../runtime-vapor/src')

export interface RootIRNode extends BaseIRNode {
  type: IRNodeTypes.ROOT
  source: string
  node: RootNode
  template: Array<TemplateFactoryIRNode | FragmentFactoryIRNode>
  dynamic: IRDynamicInfo
  effect: IREffect[]
  operation: OperationNode[]
  helpers: Set<string>
  vaporHelpers: Set<VaporHelper>
}

export interface TemplateFactoryIRNode extends BaseIRNode {
  type: IRNodeTypes.TEMPLATE_FACTORY
  template: string
}

export interface FragmentFactoryIRNode extends BaseIRNode {
  type: IRNodeTypes.FRAGMENT_FACTORY
}

export interface SetPropIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_PROP
  element: number
  key: IRExpression
  value: IRExpression
  runtimeCamelize: boolean
}

export interface SetTextIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_TEXT
  element: number
  value: IRExpression
}

export interface SetEventIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_EVENT
  element: number
  key: IRExpression
  value: IRExpression
  modifiers: {
    // modifiers for addEventListener() options, e.g. .passive & .capture
    options: string[]
    // modifiers that needs runtime guards, withKeys
    keys: string[]
    // modifiers that needs runtime guards, withModifiers
    nonKeys: string[]
  }
}

export interface SetHtmlIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_HTML
  element: number
  value: IRExpression
}

export interface CreateTextNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.CREATE_TEXT_NODE
  id: number
  value: IRExpression
}

export interface InsertNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.INSERT_NODE
  element: number | number[]
  parent: number
  anchor: number
}

export interface PrependNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.PREPEND_NODE
  elements: number[]
  parent: number
}

export interface AppendNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.APPEND_NODE
  elements: number[]
  parent: number
}

export interface WithDirectiveIRNode extends BaseIRNode {
  type: IRNodeTypes.WITH_DIRECTIVE
  element: number
  dir: VaporDirectiveNode
}

export type IRNode =
  | OperationNode
  | RootIRNode
  | TemplateFactoryIRNode
  | FragmentFactoryIRNode
export type OperationNode =
  | SetPropIRNode
  | SetTextIRNode
  | SetEventIRNode
  | SetHtmlIRNode
  | CreateTextNodeIRNode
  | InsertNodeIRNode
  | PrependNodeIRNode
  | AppendNodeIRNode
  | WithDirectiveIRNode

export interface IRDynamicInfo {
  id: number | null
  referenced: boolean
  /** created by DOM API */
  ghost: boolean
  placeholder: number | null
  children: IRDynamicChildren
}
export type IRDynamicChildren = Record<number, IRDynamicInfo>

export type IRExpression = SimpleExpressionNode | string
export interface IREffect {
  // TODO multi-expression effect
  expressions: IRExpression[]
  operations: OperationNode[]
}

type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> &
  Pick<U, Extract<keyof U, keyof T>>

export type HackOptions<T> = Prettify<
  Overwrite<
    T,
    {
      nodeTransforms?: NodeTransform[]
      directiveTransforms?: Record<string, DirectiveTransform | undefined>
    }
  >
>

export type VaporDirectiveNode = Overwrite<
  DirectiveNode,
  {
    exp: Exclude<DirectiveNode['exp'], CompoundExpressionNode>
    arg: Exclude<DirectiveNode['arg'], CompoundExpressionNode>
  }
>
