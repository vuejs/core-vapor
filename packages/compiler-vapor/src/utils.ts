import type { NumericLiteral, StringLiteral } from '@babel/types'
import { isGloballyAllowed } from '@vue/shared'
import {
  type AttributeNode,
  type ElementNode,
  NodeTypes,
  type SimpleExpressionNode,
  findDir as _findDir,
  findProp as _findProp,
  createSimpleExpression,
  isLiteralWhitelisted,
} from '@vue/compiler-dom'
import type { VaporDirectiveNode } from './ir'
import { EMPTY_EXPRESSION } from './transforms/utils'

export const findProp = _findProp as (
  node: ElementNode,
  name: string,
  dynamicOnly?: boolean,
  allowEmpty?: boolean,
) => AttributeNode | VaporDirectiveNode | undefined

/** find directive */
export const findDir = _findDir as (
  node: ElementNode,
  name: string | RegExp,
  allowEmpty?: boolean,
) => VaporDirectiveNode | undefined

export function propToExpression(prop: AttributeNode | VaporDirectiveNode) {
  return prop.type === NodeTypes.ATTRIBUTE
    ? prop.value
      ? createSimpleExpression(prop.value.content, true, prop.value.loc)
      : EMPTY_EXPRESSION
    : prop.exp
}

export function isConstantExpression(exp: SimpleExpressionNode) {
  return (
    isLiteralWhitelisted(exp.content) ||
    isGloballyAllowed(exp.content) ||
    getLiteralExpressionValue(exp) !== null
  )
}

export function resolveExpression(exp: SimpleExpressionNode) {
  if (!exp.isStatic) {
    const value = getLiteralExpressionValue(exp)
    if (value !== null) {
      return createSimpleExpression('' + value, true, exp.loc)
    }
  }
  return exp
}

export function getLiteralExpressionValue(
  exp: SimpleExpressionNode,
): number | string | null {
  if (
    !__BROWSER__ &&
    exp.ast &&
    ['StringLiteral', 'NumericLiteral'].includes(exp.ast.type)
  ) {
    return (exp.ast as StringLiteral | NumericLiteral).value
  }
  return exp.isStatic ? exp.content : null
}
