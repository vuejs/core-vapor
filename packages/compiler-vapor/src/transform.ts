import {
  type AllNode,
  type TransformOptions as BaseTransformOptions,
  type CommentNode,
  type CompilerCompatOptions,
  type ElementNode,
  ElementTypes,
  NodeTypes,
  type RootNode,
  type SimpleExpressionNode,
  type TemplateChildNode,
  defaultOnError,
  defaultOnWarn,
  isVSlot,
} from '@vue/compiler-dom'
import { EMPTY_OBJ, NOOP, extend, isArray, isString } from '@vue/shared'
import {
  type BlockIRNode,
  DynamicFlag,
  type HackOptions,
  type IRDynamicInfo,
  IRNodeTypes,
  type OperationNode,
  type RootIRNode,
  type VaporDirectiveNode,
} from './ir'
import { isConstantExpression } from './utils'
import { genDefaultDynamic } from './transforms/utils'

export type NodeTransform = (
  node: RootNode | TemplateChildNode,
  context: TransformContext<RootNode | TemplateChildNode>,
) => void | (() => void) | (() => void)[]

export type DirectiveTransform = (
  dir: VaporDirectiveNode,
  node: ElementNode,
  context: TransformContext<ElementNode>,
) => DirectiveTransformResult | void

export interface DirectiveTransformResult {
  key: SimpleExpressionNode
  value: SimpleExpressionNode
  modifier?: '.' | '^'
  runtimeCamelize?: boolean
  handler?: boolean
}

// A structural directive transform is technically also a NodeTransform;
// Only v-if and v-for fall into this category.
export type StructuralDirectiveTransform = (
  node: ElementNode,
  dir: VaporDirectiveNode,
  context: TransformContext<ElementNode>,
) => void | (() => void)

export type TransformOptions = HackOptions<BaseTransformOptions>

export interface TransformContext<T extends AllNode = AllNode> {
  node: T
  parent: TransformContext<RootNode | ElementNode> | null
  root: TransformContext<RootNode>
  index: number
  block: BlockIRNode
  options: Required<
    Omit<TransformOptions, 'filename' | keyof CompilerCompatOptions>
  >

  template: string
  childrenTemplate: (string | null)[]
  dynamic: IRDynamicInfo

  comment: CommentNode[]

  inVOnce: boolean
  inVFor: number

  enterBlock(ir: TransformContext['block'], isVFor?: boolean): () => void
  reference(): number
  increaseId(): number
  registerTemplate(): number
  registerEffect(
    expressions: SimpleExpressionNode[],
    operation: OperationNode[],
  ): void
  registerOperation(...operations: OperationNode[]): void
}

const defaultOptions = {
  filename: '',
  prefixIdentifiers: false,
  hoistStatic: false,
  hmr: false,
  cacheHandlers: false,
  nodeTransforms: [],
  directiveTransforms: {},
  transformHoist: null,
  isBuiltInComponent: NOOP,
  isCustomElement: NOOP,
  expressionPlugins: [],
  scopeId: null,
  slotted: true,
  ssr: false,
  inSSR: false,
  ssrCssVars: ``,
  bindingMetadata: EMPTY_OBJ,
  inline: false,
  isTS: false,
  onError: defaultOnError,
  onWarn: defaultOnWarn,
}

// TODO use class for better perf
function createRootContext(
  root: RootIRNode,
  node: RootNode,
  options: TransformOptions = {},
): TransformContext<RootNode> {
  let globalId = 0

  const context: TransformContext<RootNode> = {
    node,
    parent: null,
    index: 0,
    root: null!, // set later
    block: root.block,
    enterBlock(ir, inVFor = false) {
      const { block, template, dynamic, childrenTemplate } = this
      this.block = ir
      this.dynamic = ir.dynamic
      this.template = ''
      this.childrenTemplate = []
      inVFor && this.inVFor++
      return () => {
        // exit
        this.block = block
        this.template = template
        this.dynamic = dynamic
        this.childrenTemplate = childrenTemplate
        inVFor && this.inVFor--
      }
    },
    options: extend({}, defaultOptions, options),
    dynamic: root.block.dynamic,
    inVOnce: false,
    inVFor: 0,
    comment: [],

    increaseId: () => globalId++,
    reference() {
      if (this.dynamic.id !== undefined) return this.dynamic.id
      this.dynamic.flags |= DynamicFlag.REFERENCED
      return (this.dynamic.id = this.increaseId())
    },
    registerEffect(expressions, operations) {
      expressions = expressions.filter(exp => !isConstantExpression(exp))
      if (this.inVOnce || expressions.length === 0) {
        return this.registerOperation(...operations)
      }
      const existing = this.block.effect.find(e =>
        isSameExpression(e.expressions, expressions),
      )
      if (existing) {
        existing.operations.push(...operations)
      } else {
        this.block.effect.push({
          expressions,
          operations,
        })
      }

      function isSameExpression(
        a: SimpleExpressionNode[],
        b: SimpleExpressionNode[],
      ) {
        if (a.length !== b.length) return false
        return a.every((exp, i) => exp.content === b[i].content)
      }
    },

    template: '',
    childrenTemplate: [],
    registerTemplate() {
      if (!this.template) {
        return -1
      }

      const existing = root.template.findIndex(
        template => template === this.template,
      )
      if (existing !== -1) {
        return (this.dynamic.template = existing)
      }

      root.template.push(this.template)
      return (this.dynamic.template = root.template.length - 1)
    },
    registerOperation(...node) {
      this.block.operation.push(...node)
    },
  }
  context.root = context
  return context
}

// AST -> IR
export function transform(
  root: RootNode,
  options: TransformOptions = {},
): RootIRNode {
  const ir: RootIRNode = {
    type: IRNodeTypes.ROOT,
    node: root,
    source: root.source,
    template: [],
    block: {
      type: IRNodeTypes.BLOCK,
      node: root,
      dynamic: extend(genDefaultDynamic(), {
        flags: DynamicFlag.REFERENCED,
      } satisfies Partial<IRDynamicInfo>),
      effect: [],
      operation: [],
      returns: [],
    },
  }

  const context = createRootContext(ir, root, options)

  transformNode(context)

  return ir
}

export function transformNode(
  context: TransformContext<RootNode | TemplateChildNode>,
) {
  let { node } = context

  // apply transform plugins
  const { nodeTransforms } = context.options
  const exitFns = []
  for (const nodeTransform of nodeTransforms) {
    const onExit = nodeTransform(node, context)
    if (onExit) {
      if (isArray(onExit)) {
        exitFns.push(...onExit)
      } else {
        exitFns.push(onExit)
      }
    }
    if (!context.node) {
      // node was removed
      return
    } else {
      // node may have been replaced
      node = context.node
    }
  }

  // exit transforms
  context.node = node
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }

  if (context.node.type === NodeTypes.ROOT) {
    context.registerTemplate()
  }
}

export function createStructuralDirectiveTransform(
  name: string | string[],
  fn: StructuralDirectiveTransform,
): NodeTransform {
  const matches = (n: string) =>
    isString(name) ? n === name : name.includes(n)

  return (node, context) => {
    if (node.type === NodeTypes.ELEMENT) {
      const { props } = node
      // structural directive transforms are not concerned with slots
      // as they are handled separately in vSlot.ts
      if (node.tagType === ElementTypes.TEMPLATE && props.some(isVSlot)) {
        return
      }
      const exitFns = []
      for (const prop of props) {
        if (prop.type === NodeTypes.DIRECTIVE && matches(prop.name)) {
          const onExit = fn(
            node,
            prop as VaporDirectiveNode,
            context as TransformContext<ElementNode>,
          )
          if (onExit) exitFns.push(onExit)
        }
      }
      return exitFns
    }
  }
}
