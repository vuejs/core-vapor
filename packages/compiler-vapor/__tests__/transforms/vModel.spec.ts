import { makeCompile } from './_utils'
import { transformElement, transformVModel } from '../../src'
import { DOMErrorCodes } from '@vue/compiler-dom'

const compileWithVModel = makeCompile({
  nodeTransforms: [transformElement],
  directiveTransforms: {
    model: transformVModel,
  },
})

describe('compiler: vModel transform', () => {
  test('should support simple expression', () => {
    const { code, vaporHelpers } = compileWithVModel(
      '<input v-model="model" />',
    )

    expect(vaporHelpers).toContain('vModelText')
    expect(code).toMatchSnapshot()
  })

  test('should support input (text)', () => {
    const { code, vaporHelpers } = compileWithVModel(
      '<input type="text" v-model="model" />',
    )
    expect(vaporHelpers).toContain('vModelText')
    expect(code).toMatchSnapshot()
  })

  test('should support input (radio)', () => {
    const { code, vaporHelpers } = compileWithVModel(
      '<input type="radio" v-model="model" />',
    )
    expect(vaporHelpers).toContain('vModelRadio')
    expect(code).toMatchSnapshot()
  })

  test('should support input (checkbox)', () => {
    const { code, vaporHelpers } = compileWithVModel(
      '<input type="checkbox" v-model="model" />',
    )
    expect(vaporHelpers).toContain('vModelCheckbox')
    expect(code).toMatchSnapshot()
  })

  test('should support select', () => {
    const { code, vaporHelpers } = compileWithVModel(
      '<select v-model="model" />',
    )

    expect(vaporHelpers).toContain('vModelSelect')
    expect(code).toMatchSnapshot()
  })

  test('should support textarea', () => {
    const { code, vaporHelpers } = compileWithVModel(
      '<textarea v-model="model" />',
    )

    expect(vaporHelpers).toContain('vModelText')
    expect(code).toMatchSnapshot()
  })

  test('should support input (dynamic type)', () => {
    const { code, vaporHelpers } = compileWithVModel(
      '<input :type="foo" v-model="model" />',
    )
    expect(vaporHelpers).toContain('vModelDynamic')
    expect(code).toMatchSnapshot()
  })

  test('should support w/ dynamic v-bind', () => {
    const root1 = compileWithVModel('<input v-bind="obj" v-model="model" />')
    expect(root1.vaporHelpers).toContain('vModelDynamic')
    expect(root1.code).toMatchSnapshot()

    const root2 = compileWithVModel(
      '<input v-bind:[key]="val" v-model="model" />',
    )
    expect(root2.vaporHelpers).toContain('vModelDynamic')
    expect(root2.code).toMatchSnapshot()
  })

  describe('errors', () => {
    test('invalid element', () => {
      const onError = vi.fn()
      compileWithVModel('<span v-model="model" />', { onError })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: DOMErrorCodes.X_V_MODEL_ON_INVALID_ELEMENT,
        }),
      )
    })

    // TODO: component
    test.fails('plain elements with argument', () => {
      const onError = vi.fn()
      compileWithVModel('<input v-model:value="model" />', { onError })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: DOMErrorCodes.X_V_MODEL_ARG_ON_ELEMENT,
        }),
      )
    })

    // TODO: component
    test.fails('should allow usage on custom element', () => {
      const onError = vi.fn()
      const root = compileWithVModel('<my-input v-model="model" />', {
        onError,
        isCustomElement: tag => tag.startsWith('my-'),
      })
      expect(root.helpers).toContain('vModelText')
      expect(onError).not.toHaveBeenCalled()
    })

    test('should raise error if used file input element', () => {
      const onError = vi.fn()
      compileWithVModel(`<input type="file" v-model="test"/>`, {
        onError,
      })
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: DOMErrorCodes.X_V_MODEL_ON_FILE_INPUT_ELEMENT,
        }),
      )
    })

    test('should error on dynamic value binding alongside v-model', () => {
      const onError = vi.fn()
      compileWithVModel(`<input v-model="test" :value="test" />`, {
        onError,
      })
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: DOMErrorCodes.X_V_MODEL_UNNECESSARY_VALUE,
        }),
      )
    })

    // #3596
    test('should NOT error on static value binding alongside v-model', () => {
      const onError = vi.fn()
      compileWithVModel(`<input v-model="test" value="test" />`, {
        onError,
      })
      expect(onError).not.toHaveBeenCalled()
    })
  })

  describe('modifiers', () => {
    test('.number', () => {
      const { code } = compileWithVModel('<input v-model.number="model" />')

      expect(code).toMatchSnapshot()
    })

    test('.trim', () => {
      const { code } = compileWithVModel('<input v-model.trim="model" />')

      expect(code).toMatchSnapshot()
    })

    test('.lazy', () => {
      const { code } = compileWithVModel('<input v-model.lazy="model" />')

      expect(code).toMatchSnapshot()
    })
  })
})
