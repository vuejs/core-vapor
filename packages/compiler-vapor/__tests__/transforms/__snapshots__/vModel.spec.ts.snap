// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: vModel transform > modifiers > .lazy 1`] = `
"import { children as _children, vModelText as _vModelText, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelText, () => _ctx.model, void 0, { lazy: true }]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > modifiers > .number 1`] = `
"import { children as _children, vModelText as _vModelText, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelText, () => _ctx.model, void 0, { number: true }]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > modifiers > .trim 1`] = `
"import { children as _children, vModelText as _vModelText, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelText, () => _ctx.model, void 0, { trim: true }]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should support input (checkbox) 1`] = `
"import { children as _children, vModelCheckbox as _vModelCheckbox, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"checkbox\\">")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelCheckbox, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should support input (dynamic type) 1`] = `
"import { children as _children, vModelDynamic as _vModelDynamic, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelDynamic, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should support input (radio) 1`] = `
"import { children as _children, vModelRadio as _vModelRadio, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"radio\\">")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelRadio, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should support input (text) 1`] = `
"import { children as _children, vModelText as _vModelText, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"text\\">")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelText, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should support select 1`] = `
"import { children as _children, vModelSelect as _vModelSelect, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<select></select>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelSelect, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should support simple expression 1`] = `
"import { children as _children, vModelText as _vModelText, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelText, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should support textarea 1`] = `
"import { children as _children, vModelText as _vModelText, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<textarea></textarea>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelText, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should support w/ dynamic v-bind 1`] = `
"import { children as _children, vModelDynamic as _vModelDynamic, withDirectives as _withDirectives, on as _on, renderEffect as _renderEffect, setDynamicProps as _setDynamicProps, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelDynamic, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  _renderEffect(() => _setDynamicProps(n1, _ctx.obj))
  return n0
}"
`;

exports[`compiler: vModel transform > should support w/ dynamic v-bind 2`] = `
"import { children as _children, vModelDynamic as _vModelDynamic, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelDynamic, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should work with input (checkbox) 1`] = `
"import { children as _children, vModelCheckbox as _vModelCheckbox, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"checkbox\\">")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelCheckbox, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should work with input (dynamic type) 1`] = `
"import { children as _children, vModelDynamic as _vModelDynamic, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelDynamic, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should work with input (radio) 1`] = `
"import { children as _children, vModelRadio as _vModelRadio, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"radio\\">")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelRadio, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should work with input (text) 1`] = `
"import { children as _children, vModelText as _vModelText, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input type=\\"text\\">")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelText, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should work with select 1`] = `
"import { children as _children, vModelSelect as _vModelSelect, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<select></select>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelSelect, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;

exports[`compiler: vModel transform > should work with simple expression 1`] = `
"import { children as _children, vModelText as _vModelText, withDirectives as _withDirectives, on as _on, template as _template } from 'vue/vapor';
const t0 = _template("<input>")

export function render(_ctx) {
  const n0 = t0()
  const n1 = _children(n0, 0)
  _withDirectives(n1, [[_vModelText, () => _ctx.model]])
  _on(n1, "update:modelValue", $event => (_ctx.model = $event))
  return n0
}"
`;
